### ISSUE TITLE
Agents to respect .gitignore when exploring repo

### ISSUE THREAD

**Issue by 4tal** [OpenDevin/OpenDevin#2196]
## Describe your question

I want the project to ignore searching through node_modules. 

## Additional context

I mentioned it as part of the instructions and as part of the repo's readme. But still, sometimes I keep getting it to search through all the node_modules and get stuck.

**Comment by li-boxuan** [comment]
We could probably add an option to sandbox so that all files under .gitignore are invisible to the agent.

Could you please try prompting the agent to respect `.gitignore` (assuming your project is under git control)?

**Comment by 4tal** [comment]
Yeah, I can try. What will change from what I did when prompting and requesting specifically not to search within node_modules?

**Comment by 4tal** [comment]
Didn't work

**Comment by li-boxuan** [comment]
I am not too surprised about that. I think we should allow users to let agent skip everything defined in `.gitignore`.

**Comment by 4tal** [comment]
Yeah some kind of prompt wrapper that will be enforced with every action

**Comment by 4tal** [comment]
Good for both relevancy and security

**Comment by mamoodi** [comment]
@4tal the latest on main hides special folders like .git and node_modules. This isn't exactly the same as respecting the .gitignore but let us know if this helps with your node_modules issue.

**Comment by JeffKatzy** [comment]
I was looking at a fix similar [to this](https://github.com/OpenDevin/OpenDevin/commit/06a6ffcb09e3cff1f1dbd097503344cb801a376b), but saw it was reverted.  

@yufansong Can you provide some context as to why this was removed, and/or where the fix should for ignoring certain files should be made.  Is this something just do at the prompt level?

**Comment by tobitege** [comment]
What was reverted, were changes of mine to deprecated classes.

The net changes for the server-side filtering of paths is here:
https://github.com/OpenDevin/OpenDevin/blob/a97d0767e90934d0dcc82e281e8da76af220a937/opendevin/server/listen.py#L256

Personally, I'd be against using .gitignore as the source for filtering as that file can be hundreds of lines.

**Comment by jigsawlabs-student** [comment]
Is there a way for a prompt to include information the contents of a file, or specifically, to include the contents of the .gitignore?

After experimenting with the system prompt, I found that if I explicitly listed the files to ignore (eg. "ignore the main.py and index.py files as they are in the .gitignore", the agent would consistently ignore them.  However, instructions for the agent to first read the .gitignore and then ignore all files listed there was unsuccessful.

Specifically, adding this to the minimal_system_prompt did not work:
```
When issuing any commands or viewing any files or folders, first make sure these files are not in the .gitignore, and if they are in the .gitignore please ignore them.

So your first step should be to see if there is a .gitignore file, and if there is a .gitignore file read the file so that you know what files to ignore.

If you believe any action should be taken with a file or folder in the .gitignore file or you should search these files, instead indicate you are skipping them because they are in the .gitignore file.
```

But if I specified the file names it was fine, and had a nice UX of saying when it found a match but was ignoring it.  So I am looking for a way to read this content into a prompt.  Can someone point me in the right direction?  I couldn't find a similar example of incorporating workspace info in the prompt.

For the curious, here are some other things I tried but did not work: 

- Telling the system prompt to ignore contents of `.gitignore`, then separately, as a user asking opendevin to read the .gitignore.  It read the contents, but kept answering questions about the .gitignore listed files.
- Then as a user, sent another reminder to ignore contents of the .gitignore.  This didn't work.

Because of these experiments, I don't think sending OpenDevin a sequence of messages through code would work (like read the .gitignore, and then ignore those files).  It seems like the files need to be explicitly listed.

- I experimented with updating the `search_dir(search_term, dir_path='./')` skill so that it took an `exclude_files` argument.  This sometimes worked, but other times, it would use the `/api/list-files` to find the file.

**Comment by JeffKatzy** [comment]
Is there a way for a prompt to include information the contents of a file, or specifically, to include the contents of the `.gitignore`?

After experimenting with the system prompt, I found that if I explicitly listed the files to ignore (eg. "ignore the `main.py` and `index.py` files as they are in the `.gitignore`", the agent would consistently ignore them. However, instructions for the agent to first read the `.gitignore` and then ignore all files listed there was unsuccessful.

Specifically, adding this to the `minimal_system_prompt` did not work:

```
When issuing any commands or viewing any files or folders, first make sure these files are not in the .gitignore, and if they are in the .gitignore please ignore them.

So your first step should be to see if there is a .gitignore file, and if there is a .gitignore file read the file so that you know what files to ignore.

If you believe any action should be taken with a file or folder in the .gitignore file or you should search these files, instead indicate you are skipping them because they are in the .gitignore file.
```

But if I specified the file names it was fine, and had a nice UX of saying when it found a match but was ignoring it. So I am looking for a way to read this content into a prompt. Can someone point me in the right direction? I couldn't find a similar example of incorporating workspace info in the prompt.

For the curious, here are some other things I tried but did not work:

- Telling the system prompt to ignore contents of .gitignore, then separately, as a user asking opendevin to read the .gitignore. It read the contents, but kept answering questions about the .gitignore listed files.
- Then as a user, sent another reminder to ignore contents of the .gitignore. This didn't work.

Because of these experiments, I don't think sending OpenDevin a sequence of messages through code would work (like read the .gitignore, and then ignore those files). It seems like the files need to be explicitly listed.

- I experimented with updating the `search_dir(search_term, dir_path='./')` skill so that it took an `exclude_files` argument. This sometimes worked, but other times, it would use the `/api/list-files` to find the file.

**Comment by li-boxuan** [comment]
> I was looking at a fix similar [to this](https://github.com/OpenDevin/OpenDevin/commit/06a6ffcb09e3cff1f1dbd097503344cb801a376b), but saw it was reverted.

This commit was about GUI - whether we should display hidden folders on frontend.

@JeffKatzy I believe you are asking about the backend - how to make agent ignore everything recorded on `.gitignore`. I believe it is doable despite tedious. We should add a new config that says "respect .gitignore", and intercept the results from `ls` (and its variants).

**Comment by mamoodi** [comment]
@tobitege this can be closed due to this PR merging right?
https://github.com/OpenDevin/OpenDevin/pull/2621

**Comment by tobitege** [comment]
> @tobitege this can be closed due to this PR merging right? #2621

Yes, we can close this now. 👍
